PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/ui/DeviceLoginWindow.java

```diff
@@ -0,0 +1,139 @@
+import com.microsoft.intellij.ui.components.AzureDialogWrapper;
+import org.jetbrains.annotations.Nullable;
+import java.awt.Desktop;
+import java.awt.Window;
+import java.awt.event.WindowEvent;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.swing.Action;
+import javax.swing.JComponent;
+import javax.swing.JEditorPane;
+import javax.swing.JPanel;
+import javax.swing.event.HyperlinkEvent;
+
+public class DeviceLoginWindow extends AzureDialogWrapper {
+    private JPanel jPanel;
+    private JEditorPane editorPanel;
+    private AuthenticationResult authenticationResult = null;
+    private Future<?> authExecutor;
+    private static final String TITLE = "Azure Device Login";
+
+    public AuthenticationResult getAuthenticationResult() {
+        return authenticationResult;
+    }
+
+    public DeviceLoginWindow(final AuthenticationContext ctx, final DeviceCode deviceCode,
+                             final AuthenticationCallback<AuthenticationResult> callBack) {
+        super(null, false, IdeModalityType.PROJECT);
+        setModal(true);
+        setTitle(TITLE);
+        editorPanel.setText(createHtmlFormatMessage(deviceCode.getMessage()));
+        editorPanel.addHyperlinkListener(e -> {
+            if (e.getEventType() == HyperlinkEvent.EventType.ACTIVATED) {
+                try {
+                    Desktop.getDesktop().browse(e.getURL().toURI());
+                } catch (IOException | URISyntaxException e1) {
+                    // swallow exceptions
+                }
+            }
+        });
+        authExecutor = ApplicationManager.getApplication()
+            .executeOnPooledThread(() -> pullAuthenticationResult(ctx, deviceCode, callBack));
+        init();
+    }
+
+    private void pullAuthenticationResult(final AuthenticationContext ctx, final DeviceCode deviceCode,
+                                          final AuthenticationCallback<AuthenticationResult> callback) {
+        final long interval = deviceCode.getInterval();
+        long remaining = deviceCode.getExpiresIn();
+        while (remaining > 0 && authenticationResult == null) {
+            try {
+                remaining -= interval;
+                Thread.sleep(interval * 1000);
+                authenticationResult = ctx.acquireTokenByDeviceCode(deviceCode, callback).get();
+            } catch (ExecutionException | InterruptedException e) {
+                if (e.getCause() instanceof AuthenticationException &&
+                    ((AuthenticationException) e.getCause()).getErrorCode() == AdalErrorCode.AUTHORIZATION_PENDING) {
+                    // swallow the pending exception
+                } else {
+                    e.printStackTrace();
+                    break;
+                }
+            }
+        }
+        closeDialog();
```

Comment:
Can we just invoking `close(0)`? If it works, the method `closeDialog()` can be removed.